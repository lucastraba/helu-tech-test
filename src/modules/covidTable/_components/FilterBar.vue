<template>
  <div class="row">
    <div class="col-6">
      <b-form-group
        v-if="!isInGraphView"
        class="mb-0 filter-field"
      >
        <b-input-group size="sm">
          <b-form-input
            :value="ghostSuggestion"
            type="text"
            class="input ghost-complete"
            tabindex="-1"
            disabled
            placeholder="Filter by country"
          />
          <b-form-input
            v-model="filter"
            type="search"
            class="main-input"
            placeholder="Filter by country"
            @keyup.enter="fillSuggestion"
            @keydown.tab="fillSuggestion"
          />
          <b-input-group-append>
            <b-button
              class="clear-filter"
              :disabled="!filter"
              @click="clearFilter"
            >
              Clear
            </b-button>
          </b-input-group-append>
        </b-input-group>
      </b-form-group>
    </div>
    <div class="col d-inline-flex flex-row-reverse">
      <b-button
        class="toggle-view"
        variant="success"
        @click="toggleView"
      >
        {{ viewLabel }}
      </b-button>
    </div>
  </div>
</template>

<script>
import { TABLE_VIEWS } from '@/modules/covidTable/constants/tableConstants';

export default {
  name: 'FilterBar',
  props: {
    isInGraphView: {
      type: Boolean,
      required: true
    },
    suggestion: {
      type: String,
      required: false,
      default: ''
    }
  },
  data () {
    return {
      filter: ''
    };
  },
  computed: {
    viewLabel () {
      return (this.isInGraphView) ? TABLE_VIEWS[0] : TABLE_VIEWS[1];
    },
    ghostSuggestion () {
      if (this.filter) {
        return this.filter.substr(0, this.filter.length) + this.suggestion.substr(this.filter.length, this.suggestion.length);
      }
      return '';
    }
  },
  watch: {
    filter (value) {
      this.$emit('filtered', value);
    }
  },
  methods: {
    toggleView () {
      this.$emit('toggle-view');
    },
    fillSuggestion () {
      this.filter = this.suggestion;
    },
    clearFilter () {
      this.filter = '';
    }
  }
};
</script>
<style scoped lang="scss">
.main-input, .main-input:focus {
  background-color: transparent;
}

.ghost-complete {
  position: absolute !important;
  color: #b4b4b4;
  background-color: white !important;
  width: 100% !important;
  right: 1px;
}
</style>
