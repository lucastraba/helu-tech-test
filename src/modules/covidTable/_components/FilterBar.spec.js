import { createLocalVue, shallowMount } from '@vue/test-utils';
import FilterBar from './FilterBar';

describe('FilterBar Component', () => {
  const mock$emit = jest.fn();
  const localVue = createLocalVue();
  const factory = ({ propsData } = {}) => shallowMount(FilterBar, {
    localVue,
    stubs: {
      bFormGroup: true,
      bInputGroup: true,
      bFormInput: true,
      bButton: true,
      bInputGroupAppend: true
    },
    mocks: {
      $emit: mock$emit
    },
    propsData: {
      isInGraphView: false,
      ...propsData
    }
  });
  beforeEach(() => {
    jest.clearAllMocks();
  });
  describe('base rendering', () => {
    it('should render the filter if not in graph view and button should show graph view', () => {
      // Arrange.
      const wrapper = factory();
      // Act.
      // Assert.
      expect(wrapper.find('.filter-field').exists()).toBe(true);
      expect(wrapper.find('.clear-filter').exists()).toBe(true);
      expect(wrapper.find('.toggle-view').text()).toBe('Show Graph');
      expect(wrapper.html()).toMatchSnapshot();
    });
    it('should not render the filter if in graph view and button should show table view', () => {
      // Arrange.
      const wrapper = factory({
        propsData: {
          isInGraphView: true
        }
      });
      // Act.
      // Assert.
      expect(wrapper.find('.filter-field').exists()).toBe(false);
      expect(wrapper.find('.clear-filter').exists()).toBe(false);
      expect(wrapper.find('.toggle-view').text()).toBe('Show Table');
      expect(wrapper.html()).toMatchSnapshot();
    });
  });
  describe('computed', () => {
    describe('ghostSuggestion', () => {
      it('should return an adapted ghost suggestion', async () => {
        // Arrange.
        const wrapper = factory({
          propsData: {
            suggestion: 'Foobar'
          }
        });
        // Act.
        await wrapper.setData({
          filter: 'fo'
        });

        // Assert.
        expect(wrapper.vm.ghostSuggestion).toBe('foobar');
      });
    });
  });
  describe('watchers', () => {
    describe('filter', () => {
      it('should emit the filter to parent component', async () => {
        // Arrange.
        const wrapper = factory();
        // Act.
        await wrapper.setData({
          filter: 'foo'
        });
        // Assert.
        expect(mock$emit).toHaveBeenCalledWith('filtered', 'foo');

        await wrapper.setData({
          filter: 'bar'
        });
        expect(mock$emit).toHaveBeenCalledTimes(2);
      });
    });
  });
  describe('methods', () => {
    describe('toggleView', () => {
      it('should emit the toggle-view event to parent', async () => {
        // Arrange.
        const wrapper = factory();
        // Act.
        wrapper.vm.toggleView();
        // Assert.
        expect(mock$emit).toHaveBeenLastCalledWith('toggle-view');
      });
    });
    describe('fillSuggestion', () => {
      it('should update the filter with the suggestion', async () => {
        // Arrange.
        const wrapper = factory({
          propsData: {
            suggestion: 'Barfoobaz'
          }
        });
        // Act.
        await wrapper.setData({
          filter: 'bar'
        });
        wrapper.vm.fillSuggestion();
        // Assert.
        expect(wrapper.vm.filter).toBe('Barfoobaz');
      });
    });
    describe('clearFilter', () => {
      it('should clear the filter', async () => {
        // Arrange.
        const wrapper = factory();
        // Act.
        await wrapper.setData({
          filter: 'foo'
        });
        wrapper.vm.clearFilter();
        // Assert.
        expect(wrapper.vm.filter).toBe('');
      });
    });
  });
});
