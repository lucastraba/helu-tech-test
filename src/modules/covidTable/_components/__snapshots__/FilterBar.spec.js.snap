// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FilterBar Component base rendering should not render the filter if in graph view and button should show table view 1`] = `
"<div class=\\"row\\">
  <div class=\\"col-6\\">
    <!---->
  </div>
  <div class=\\"col d-inline-flex flex-row-reverse\\">
    <bbutton-stub variant=\\"success\\" class=\\"toggle-view\\">
      Show Table
    </bbutton-stub>
  </div>
</div>"
`;

exports[`FilterBar Component base rendering should render the filter if not in graph view and button should show graph view 1`] = `
"<div class=\\"row\\">
  <div class=\\"col-6\\">
    <bformgroup-stub class=\\"mb-0 filter-field\\">
      <binputgroup-stub size=\\"sm\\">
        <bforminput-stub value=\\"\\" type=\\"text\\" tabindex=\\"-1\\" disabled=\\"\\" placeholder=\\"Filter by country\\" class=\\"input ghost-complete\\"></bforminput-stub>
        <bforminput-stub type=\\"search\\" placeholder=\\"Filter by country\\" value=\\"\\" class=\\"main-input\\"></bforminput-stub>
        <binputgroupappend-stub>
          <bbutton-stub disabled=\\"true\\" class=\\"clear-filter\\">
            Clear
          </bbutton-stub>
        </binputgroupappend-stub>
      </binputgroup-stub>
    </bformgroup-stub>
  </div>
  <div class=\\"col d-inline-flex flex-row-reverse\\">
    <bbutton-stub variant=\\"success\\" class=\\"toggle-view\\">
      Show Graph
    </bbutton-stub>
  </div>
</div>"
`;
